### PR TITLE
feat(JOML): migrate SimpleFarming

### PR DESCRIPTION
**Pull request by pollend**
- clean up code
- fix test cases
PR: Terasology/ModuleTestingEnvironment#27

### CODE REVIEW

File: src/main/java/org/terasology/simpleFarming/events/OnSeedPlanted.java [hunk 1]
```diff
@@ -1,22 +1,9 @@
...
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
 package org.terasology.simpleFarming.events;
 
+import org.joml.Vector3i;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/simpleFarming/events/OnSeedPlanted.java:5]: This change is straight forward, but is the resulting class the way we want it to be? It hands out it's internal position vector via `getPosition` - mutating this outside can have weird consequences.

https://github.com/Terasology/SimpleFarming/blob/e6102d983a88fa5682ece68ca202cbdb903786b4/src/main/java/org/terasology/simpleFarming/events/OnSeedPlanted.java#L24-L26

Should these getters only return the interfaces/constants (`Vector3ic`) instead?

File: src/test/java/org/terasology/simpleFarming/systems/PlantAuthoritySystemTest.java [hunk 2]
```diff
@@ -35,135 +23,98 @@
...
-    @Override
-    public Set<String> getDependencies() {
-        return Collections.singleton("SimpleFarming");
-    }
-
     @Test
-    @Ignore // TODO Seems to enter infinite loop
     public void seedShouldGrowWherePlanted() {
-        plantSeedTest(dirt, air, true);
+        setBlock(new Vector3i(), dirt);
```
Comment (skaldarnar) [inline src/test/java/org/terasology/simpleFarming/systems/PlantAuthoritySystemTest.java]: I'd prefer to be explicit here in the test with the vector coordinates.
```suggestion
        setBlock(new Vector3i(0, 0, 0), dirt);
```

File: src/test/java/org/terasology/simpleFarming/systems/PlantAuthoritySystemTest.java [hunk 3]
```diff
@@ -35,135 +23,98 @@
...
+        setBlock(new Vector3i(0, 1, 0), air);
+
+        EntityRef seed = testSeed();
+        plant(seed, new Vector3f());
```
Comment (skaldarnar) [inline src/test/java/org/terasology/simpleFarming/systems/PlantAuthoritySystemTest.java]:
```suggestion
        plant(seed, new Vector3f(0, 0, 0));
```

File: src/test/java/org/terasology/simpleFarming/systems/PlantAuthoritySystemTest.java [hunk 4]
```diff
@@ -35,135 +23,98 @@
...
+
+        Assertions.assertEquals(bush, worldProvider.getBlock(new Vector3i(0, 1, 0)));
+        Assertions.assertEquals(dirt, worldProvider.getBlock(new Vector3i(0, 0, 0)));
     }
 
     @Test
-    @Ignore // TODO Seems to enter infinite loop
-    public void seedShouldNotGrowOnAir() {
-        plantSeedTest(air, air, false);
+    public void tryPlanSeedOnAir() {
```
Comment (skaldarnar) [inline src/test/java/org/terasology/simpleFarming/systems/PlantAuthoritySystemTest.java]: typo
```suggestion
    public void tryPlantSeedOnAir() {
```

File: src/test/java/org/terasology/simpleFarming/systems/BushAuthoritySystemTest.java [hunk 5]
```diff
@@ -1,57 +1,162 @@
...
+
+        // harvest bush
+        final TestEventReceiver<DropItemEvent> dropSpy = new TestEventReceiver<>(helper.getHostContext(), DropItemEvent.class);
+        entity.send(new ActivateEvent(entity, playerInstigator(), null, null, null, null, 0));
+
+        // check if bush drops produce
+        Assertions.assertTrue(dropSpy.getEntityRefs().stream().anyMatch(k -> k.getParentPrefab().getName().equals(component.produce)));
+
+        // check if bush is at max growth state
+        assertEquals(component.growthStages.size() - 2, component.currentStage);
```
Comment (skaldarnar) [inline src/test/java/org/terasology/simpleFarming/systems/BushAuthoritySystemTest.java:98]: So, the max growth stage is one state before the final growth state? This is confusing :roll_eyes:

File: src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java [hunk 6]
```diff
@@ -306,8 +294,8 @@ private boolean isValidPosition(Vector3i position) {
         if (targetBlock != airBlock) {
             return false;
         }
-        Block belowBlock = worldProvider.getBlock(position.addY(-1));
-        position.addY(1);
+        Block belowBlock = worldProvider.getBlock(position.add(0,-1,0));
+        position.add(0,1,0);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/simpleFarming/systems/VineAuthoritySystem.java:298]: I can already see how error prone this is :see_no_evil:

File: src/main/java/org/terasology/simpleFarming/systems/PlantAuthoritySystem.java [hunk 7]
```diff
@@ -122,9 +109,9 @@ private boolean isValidPosition(Vector3i position) {
         Block targetBlock = worldProvider.getBlock(position);
 
         /* Avoid construction of a transient Vector3i in order to save memory */
-        position.addY(-1);
+        position.y += -1;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/simpleFarming/systems/PlantAuthoritySystem.java]: Why not 
```suggestion
        position.y -= 1;
```

File: src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java [hunk 8]
```diff
@@ -298,11 +288,11 @@ public void onPlantDestroyed(DoDestroyPlant event, EntityRef entity, BushDefinit
      *
      * @param bushComponent the bush component of the entity
      */
-    private void onBushDestroyed(Vector3i position, EntityRef bush, BushDefinitionComponent bushComponent) {
+    private void onBushDestroyed(Vector3ic position, EntityRef bush, BushDefinitionComponent bushComponent) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/simpleFarming/systems/BushAuthoritySystem.java:291]: :+1:

**Review by skaldarnar — COMMENTED** [review]
Here, have a few comments on this one :upside_down_face: Overall looks fine, I'm just wondering how much effort we should put into setting all the interfaces straight to an idiomatic use of JOML…?

**Review by jdrueckert — APPROVED** [review]

**Pull request merged**